libraries/chain/asset_evaluator.cpp

```diff
@@ -360,42 +360,194 @@ void_result asset_update_issuer_evaluator::do_apply(const asset_update_issuer_op
-   FC_ASSERT( !b.has_settlement(), "Cannot update a bitasset after a settlement has executed" );
-   if( o.new_options.short_backing_asset != b.options.short_backing_asset )
+   FC_ASSERT( op.issuer == asset_obj.issuer, "Only asset issuer can update bitasset_data of the asset." );
+
+   const asset_bitasset_data_object& current_bitasset_data = asset_obj.bitasset_data(d);
+
+   FC_ASSERT( !current_bitasset_data.has_settlement(), "Cannot update a bitasset after a global settlement has executed" );
+
+   bool after_hf_core_922_931 = ( d.get_dynamic_global_properties().next_maintenance_time > HARDFORK_CORE_922_931_TIME );
+
+   // Are we changing the backing asset?
+   if( op.new_options.short_backing_asset != current_bitasset_data.options.short_backing_asset )
    {
-      FC_ASSERT(a.dynamic_asset_data_id(d).current_supply == 0);
-      FC_ASSERT(d.find_object(o.new_options.short_backing_asset));
+      FC_ASSERT( asset_obj.dynamic_asset_data_id(d).current_supply == 0,
+                 "Cannot update a bitasset if there is already a current supply." );
+
+      const asset_object& new_backing_asset = op.new_options.short_backing_asset(d); // check if the asset exists
+
+      if( after_hf_core_922_931 ) // TODO remove this check after hard fork if things in `else` did not occur
+      {
+         FC_ASSERT( op.new_options.short_backing_asset != asset_obj.get_id(),
+                    "Cannot update an asset to be backed by itself." );
+
+         if( current_bitasset_data.is_prediction_market )
+         {
+            FC_ASSERT( asset_obj.precision == new_backing_asset.precision,
+                       "The precision of the asset and backing asset must be equal." );
+         }
+
+         if( asset_obj.issuer == GRAPHENE_COMMITTEE_ACCOUNT )
+         {
+            if( new_backing_asset.is_market_issued() )
+            {
+               FC_ASSERT( new_backing_asset.bitasset_data(d).options.short_backing_asset == asset_id_type(),
+                          "May not modify a blockchain-controlled market asset to be backed by an asset which is not "
+                          "backed by CORE." );
+            }
+            else
+            {
+               FC_ASSERT( new_backing_asset.get_id() == asset_id_type(),
+                          "May not modify a blockchain-controlled market asset to be backed by an asset which is not "
+                          "market issued asset nor CORE." );
+            }
+         }
+         else
+         {
+            // not a committee issued asset
+
+            // If we're changing to a backing_asset that is not CORE, we need to look at any
+            // asset ( "CHILD" ) that has this one as a backing asset. If CHILD is committee-owned,
+            // the change is not allowed. If CHILD is user-owned, then this asset's backing
+            // asset must be either CORE or a UIA.
+            if ( new_backing_asset.get_id() != asset_id_type() ) // not backed by CORE
+            {
```

Comment:
Still need to check `new_backing_asset` here, to see whether this will result in an MPA (this asset) backed by an MPA backed by an MPA.

This should be done regardless of CHILD iteration.

Scenario: there was asset B backed by asset C backed by CORE or an UIA, now we're changing C's backing asset to B, or changing A's backing asset to B, neither should be allowed.

Reply:
Refactored (as suggested below), added check for A backed by B backed by A

Reply:
Still need to check for A backed by B backed by C (non-CORE), when this operation is updating A.

Reply:
Added a check to make sure C is not market issued when A is changing B.